### PR TITLE
Autocomplete: Add option to stream code completion results

### PR DESCRIPTION
**Pull request by philipp-spiess**
This adds a new boolean `stream` to the code completion handler. When set, the resulting response will be an SSE stream instead of requiring us to buffer the response.

This goes hand-in-hand with these client changes: https://github.com/sourcegraph/cody/pull/723 however, both PRs can be implemented in parallel and are backward compatible. If a client sends the stream flag but a regular HTTP response (as detected by the response header) is encountered, it'll fall back to the old JSON warning. If a server gets a response with `stream: true` that does not have this change, it will response with the previous format.

Note: There's a lot of repetition here. If someone could help me with more idiomatic Go patters for this fork I'd highly appreciate it 😬 

## Test plan

Tested in combination with https://github.com/sourcegraph/cody/pull/723

 - Run [sourcegraph/sourcegraph#55967](https://github.com/sourcegraph/sourcegraph/pull/55967) as a local SG server
 - Set up Cody VS Code to connect to localhost
 - Check out https://github.com/sourcegraph/cody/pull/723
 - Start a dev instance and see that completions will work and might be faster
   - I've tested this with our completion testing tool but since we only have a handful of examples and don't do a bazillion test runs, there's a log of variance in the early results. 
   - We can roll this out as an A/B test very easily (cc @taras-yemets)

### CODE REVIEW

File: internal/completions/types/types.go [hunk 1]
```diff
@@ -58,6 +59,7 @@ type CompletionRequestParameters struct {
 	TopK              int       `json:"topK,omitempty"`
 	TopP              float32   `json:"topP,omitempty"`
 	Model             string    `json:"model,omitempty"`
+	Stream            bool      `json:"stream,omitempty"`
```
Comment (eseliger) [inline internal/completions/types/types.go:62]: This struct is shared among chat and code completions, so the property will actually have no effect for chat rn. Although it has no effect, it's value will be `false` by default, which I think can be misleading.

File: internal/completions/httpapi/codecompletion.go [hunk 2]
```diff
@@ -24,32 +25,74 @@ func NewCodeCompletionsHandler(logger log.Logger, db database.DB) http.Handler {
...
-			logFields := []log.Field{log.Error(err)}
-
-			// Propagate the upstream headers to the client if available.
-			if errNotOK, ok := types.IsErrStatusNotOK(err); ok {
-				errNotOK.WriteHeader(w)
-				if tc := errNotOK.SourceTraceContext; tc != nil {
-					logFields = append(logFields,
-						log.String("sourceTraceContext.traceID", tc.TraceID),
-						log.String("sourceTraceContext.spanID", tc.SpanID))
+		if requestParams.Stream {
```
Comment (eseliger) [inline internal/completions/httpapi/codecompletion.go:28]: this is basically duplicating all the code in the response handler for the chat handler, I opened another PR here that uses a bit less of duplication, LMK what you think!

https://github.com/sourcegraph/sourcegraph/pull/55977

**Review by eseliger — COMMENTED** [review]
This works and if you need to move super fast here, feel free to merge this PR now and we can follow up on my comments, but I left one thing about Stream being false by default for chat, and proposed a slightly different solution with less code duplication in a draft :)

**Comment by philipp-spiess** [comment]
@eseliger 

step 1: tell erik about a change i want to make
step 2: write 💩  go code 
step 3: wait for erik to fix it for me 😍 

I like your PR much more, let's move forward with that instead.

**Comment by eseliger** [comment]
Lol haha works for me, could you test that one locally to make sure nothing broke? 😇 our unit tests are.. well there ain’t some

**Comment by philipp-spiess** [comment]
Focusing on #55977 instead

**Pull request closed without merging**